Adding iterative batch function for IPs

Samuel-Gill opened the pull request:
In this PR, I have added the iterative function to get the IP details 1 by 1.

UmanShahzad reviewed (changes requested):
Missing caching, error handling and detail formatting. See other funcs carefully

Samuel-Gill commented on ipinfo/handler.py:
```diff
@@ -305,3 +305,24 @@ def getMap(self, ips):
+        headers["content-type"] = "application/json"
+
+        # Split the IP addresses into batches
+        batches = [
```
Updated.

Samuel-Gill commented:
> Missing caching, error handling and detail formatting. See other funcs carefully

I've updated the suggested changes, please review.

UmanShahzad reviewed (changes requested):
This new implementation isn't even doing what the point of the iterator is - it's also just collecting all the data and at the end doing a yield.

We also need a version of this in `handler_async` that uses async features.

Samuel-Gill commented:
> This new implementation isn't even doing what the point of the iterator is - it's also just collecting all the data and at the end doing a yield.
> 
> We also need a version of this in `handler_async` that uses async features.

I have updated the handler function it now directly yields the details obtained from the batch lookup API call. This way, the function will yield the response after each batch lookup iteration, without collecting all the data and yielding at the end.

Also added a version in `handler_async` . Please review @UmanShahzad

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -305,3 +305,62 @@ def getMap(self, ips):
+            try:
+                response = requests.post(url, json=batch, headers=headers)
+            except Exception as e:
+                return handler_utils.return_or_fail(raise_on_fail, e)
```
Dont think we can return here? How does it behave in an iterator in python? 

See https://stackoverflow.com/questions/26595895/return-and-yield-in-the-same-function

Samuel-Gill replied:
Updated to raise exception.

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -305,3 +305,62 @@ def getMap(self, ips):
+            )
+            for ip in batch:
+                detail = details.get(ip)
+                if detail is not None:
```
In what situation would you get `detail is None`?

Samuel-Gill replied:
Updated.

UmanShahzad commented on tests/handler_test.py:
```diff
@@ -128,6 +131,19 @@ def _check_batch_details(ips, details, token):
             assert "domains" in d
 
 
+def _check_iterative_batch_details(details, token):
```
Can we just use `_check_batch_details`? Just pass in a single IP array containing the returned detail's IP

Samuel-Gill replied:
No, actually `getBatchDetails` returns `dict` as we can simply loop through the entries and check the details. On the other hand `getBatchDetailsIter` is returning generator object and we yield them iteratively. I tried storing it in separate variable first before yield, but it gets override by new value so it arrives. Also we have to do some quirks to get the detail value as it is not subscriptable.

UmanShahzad commented on tests/handler_async_test.py:
```diff
@@ -142,6 +145,31 @@ async def test_get_batch_details(batch_size):
     await handler.deinit()
 
 
+def _check_iterative_batch_details(ip, details, token):
```
Same here as for sync test comment above

UmanShahzad commented on tests/handler_async_test.py:
```diff
@@ -142,6 +145,31 @@ async def test_get_batch_details(batch_size):
+    async for ips, details in handler.getIterativeBatchDetails(
+        ips, batch_size
+    ):
+        # results.append((ip_address, details))
```
?

Samuel-Gill replied:
Removed.

UmanShahzad commented on tests/handler_test.py:
```diff
@@ -105,7 +108,7 @@ def test_get_details():
 def _prepare_batch_test():
     """Helper for preparing batch test cases."""
-    token = os.environ.get("IPINFO_TOKEN", "")
+    token =  os.environ.get("IPINFO_TOKEN", "")
```
Why extra space?

Samuel-Gill replied:
Updated.

UmanShahzad commented on tests/handler_async_test.py:
```diff
@@ -142,6 +145,31 @@ async def test_get_batch_details(batch_size):
+    ):
+        # results.append((ip_address, details))
+        _check_iterative_batch_details(ips, details, token)
+        await handler.deinit()
```
How did this test case pass? This is deinitializing a handler that will get used again on the next for-loop iteration.

Samuel-Gill replied:
Updated.

UmanShahzad commented on ipinfo/handler_async.py:
```diff
@@ -347,3 +347,69 @@ def _ensure_aiohttp_ready(self):
         timeout = aiohttp.ClientTimeout(total=self.request_options["timeout"])
         self.httpsess = aiohttp.ClientSession(timeout=timeout)
+
+    async def getIterativeBatchDetails(
```
Same name comment

Samuel-Gill replied:
Updated.

UmanShahzad commented on ipinfo/handler_async.py:
```diff
@@ -347,3 +347,69 @@ def _ensure_aiohttp_ready(self):
+                    for ip_address, details in json_response.items():
+                        self.cache[cache_key(ip_address)] = details
+                        results[ip_address] = details
+
```
Unnecessary newline

Samuel-Gill replied:
Removed.

UmanShahzad commented on ipinfo/handler_async.py:
```diff
@@ -347,3 +347,69 @@ def _ensure_aiohttp_ready(self):
+                        results[ip_address] = details
+
+            except Exception as e:
+                return handler_utils.return_or_fail(raise_on_fail, e, results)
```
Same `return` comment as sync - maybe

Samuel-Gill replied:
Updated.

UmanShahzad commented on ipinfo/handler_async.py:
```diff
@@ -347,3 +347,69 @@ def _ensure_aiohttp_ready(self):
+        ip_addresses,
+        batch_size=None,
+        raise_on_fail=True,
+    ):
```
Is this code handling everything the same way the other `getBatchDetails` was? It looks like it might be missing stuff at a glance

Samuel-Gill replied:
yes cache check was missing in one, it is now added.

UmanShahzad reviewed (commented):
Let's rename the funcs to be `getBatchDetailsIter` so that the non-iter version has its name as a prefix here - easier to read/understand at a glance..

This is also getting a bit too hard to review because the code can be very subtlely wrong, so please don't ask for review until you've done your own several review passes.

Samuel-Gill commented on ipinfo/handler.py:
```diff
@@ -305,3 +305,62 @@ def getMap(self, ips):
+        if batch_size is None:
+            batch_size = BATCH_MAX_SIZE
+
+        lookup_addresses = []
```
Updated both iter and non-iter to do cache lookups and bogon check.

Samuel-Gill commented on ipinfo/handler.py:
```diff
@@ -305,3 +305,62 @@ def getMap(self, ips):
+                details["bogon"] = True
+                yield Details(details)
+            else:
+                lookup_addresses.append(ip_address)
```
Added check

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -210,6 +210,14 @@ def getBatchDetails(
+                details = {}
+                details["ip"] = ip_address
+                details["bogon"] = True
+                return Details(details)
```
This is clearly broken.

Samuel-Gill replied:
Can give me little more information about this, it will help me identify and analyze the cause.

UmanShahzad replied:
It is returning the details

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -305,3 +313,72 @@ def getMap(self, ips):
+
+        # all in cache - exit early.
+        if len(lookup_addresses) == 0:
+            yield result.items()
```
This is not exiting. It'll continue executing the function on the next loop?

Samuel-Gill replied:
I've updated, please review if it fixed it now.

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -218,7 +218,7 @@ def getBatchDetails(
                 details["ip"] = ip_address
                 details["bogon"] = True
-                return Details(details)
+                result[ip_address] = details
```
Doesn't it need to be wrapped in `Details`, or not?

Samuel-Gill replied:
Could be both `Details` just create instance and return them.

UmanShahzad commented on ipinfo/handler.py:
```diff
@@ -339,7 +339,7 @@ def getBatchDetailsIter(
                 details["ip"] = ip_address
                 details["bogon"] = True
-                yield Details(details)
+                result[ip_address] = details
```
This was correct before; it must yield.

Samuel-Gill replied:
updated.
